### ISSUE TITLE
CTCLoss: fuse the forward-backward loop to reduce per-timestep kernel dispatch overhead

### ISSUE THREAD

**Issue by antimora** [tracel-ai/burn#4817]
## Summary

The current `CTCLoss` implementation in `burn-nn/src/loss/ctc.rs` loops over timesteps sequentially (line 154: `for t in 1..max_input_length`), issuing ~40 small tensor operations per iteration (two `right_shift_2d_tensor` calls, two `log_sum_exp` calls with ~12 ops each, a `gather`, `add`, and `mask_where`). For typical ASR sequence lengths (T = 1000-2000 frames), this results in ~40,000-80,000 individual kernel dispatches per loss call.

On GPU backends (wgpu/cubecl, LibTorch-MPS), each dispatch has fixed overhead from CPU-side submission and GPU-side scheduling. The sequential dependency (timestep `t` depends on `t-1`) prevents the backend from batching or overlapping these kernels. In profiling, this manifests as oscillating CPU/GPU utilization during the loss computation — the GPU is underutilized because it spends time waiting between tiny kernels rather than doing sustained compute.

## Current behavior

Each CTC forward pass on a batch with `max_input_length = T` executes roughly:
- `T` iterations of the alpha recursion
- Per iteration: 2x `log_sum_exp` (each ~12 tensor ops for the NaN-safe path), 2x `right_shift_2d_tensor` (slice + cat), 1x `gather`, 1x `add`, 1x `mask_where`
- Total: ~40T kernel dispatches

The `log_sum_exp` helper (line 402) is particularly heavy — it handles `-inf` entries safely via masking, which requires ~12 ops per call. For the common case where most entries are finite, a simpler `max + log(1 + exp(diff))` with a single `-inf` guard would halve the op count.

### 1. Simplify `log_sum_exp` (low effort, ~2x fewer ops per timestep)

The current implementation creates multiple intermediate mask tensors to handle `-inf` entries. A more compact formulation:

```rust
fn log_sum_exp(a, b) -> Tensor {
    let max_val = a.clone().max_pair(b.clone());
    let sum = (a - max_val.clone()).exp() + (b - max_val.clone()).exp();
    max_val + sum.log()
}
```

This produces `-inf + log(0+0) = -inf` naturally when both inputs are `-inf`, without explicit masking. For the mixed case (`-inf` + finite), `exp(-inf - finite) = 0`, so the result is `finite + log(0 + 1) = finite`. Correct in all cases, ~4 ops instead of ~12.

### 2. Fused cubecl kernel for the alpha recursion (high effort, eliminates the loop)

The forward-backward algorithm has a well-known parallel structure along the label axis (each `s` position can be computed independently given `t-1`). A single cubecl kernel that loops over `t` internally — reading `alpha[t-1]` from shared memory and writing `alpha[t]` — would replace ~40T host-side dispatches with a single kernel launch. This is what PyTorch's native `CTCLoss` CUDA kernel does.

This would be a larger contribution but would make Burn's CTC loss competitive with PyTorch's for real ASR workloads (T > 500).

## Environment

- burn 0.21.0-pre.3 (git rev b62be4dd)
- Observed on both wgpu (Metal/cubecl) and LibTorch-MPS backends

**Comment by softmaximalist** [comment]
I can work on this (optimizing the existing high level tensor ops).

**Comment by antimora** [comment]
> I can work on this (optimizing the existing high level tensor ops).

Actually I was about to submit a PR

**Comment by softmaximalist** [comment]
> Actually I was about to submit a PR

No worries. I will look into your PR to learn more!

**Comment by laggui** [comment]
We can allow specialization for such use cases, but if we do that it also needs to implement the custom backward. It will no longer be differentiated automatically because it's no longer composed of tracked tensor operations.

For other custom losses, users can apply a similar pattern can be implemented via a custom backend extension. In this case it's a loss that we actually implement so it's in the backend ops instead of an extension.
